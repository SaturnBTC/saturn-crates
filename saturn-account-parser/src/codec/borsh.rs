use arch_program::account::AccountInfo;
use arch_program::program_error::ProgramError;
use borsh::{BorshDeserialize, BorshSerialize};
use std::io::Cursor;

/// Classical Borsh serialisation codec.
pub struct BorshCodec;

impl BorshCodec {
    /// Deserialises the shard held in `account` using Borsh.
    pub fn load<S>(account: &AccountInfo<'_>) -> Result<S, ProgramError>
    where
        S: BorshDeserialize + BorshSerialize,
    {
        let data = account.try_borrow_data()?;
        S::try_from_slice(&data).map_err(|_| ProgramError::InvalidAccountData)
    }

    /// Serialises `shard` back into `account.data` using Borsh.
    pub fn store<S>(account: &AccountInfo<'_>, shard: &S) -> Result<(), ProgramError>
    where
        S: BorshSerialize,
    {
        let mut data = account.try_borrow_mut_data()?;
        // Write directly into the account's data buffer via an in-place cursor.
        let mut cursor = {
            let slice: &mut [u8] = &mut *data;
            Cursor::new(slice)
        };
        shard
            .serialize(&mut cursor)
            .map_err(|_| ProgramError::InvalidAccountData)?;
        Ok(())
    }
}

/// Anchor-style Borsh account wrapper mirroring Anchor's `Account<'info, T>`.
#[allow(clippy::module_name_repetitions)]
pub struct Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    account: &'a AccountInfo<'a>,
    data: T,
}

impl<'a, T> Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    pub fn load(account: &'a AccountInfo<'a>) -> Result<Self, ProgramError> {
        let data = BorshCodec::load::<T>(account)?;
        Ok(Self { account, data })
    }

    /// Returns a reference to the underlying `AccountInfo` object.
    pub fn info(&self) -> &'a AccountInfo<'a> {
        self.account
    }

    /// Convenience helper that clones the underlying `AccountInfo`.
    pub fn clone_account(&self) -> AccountInfo<'a> {
        self.account.clone()
    }
}

impl<'a, T> Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize + Default,
{
    pub fn init(account: &'a AccountInfo<'a>) -> Result<Self, ProgramError> {
        let default_val: T = Default::default();
        BorshCodec::store::<T>(account, &default_val)?;
        Ok(Self {
            account,
            data: default_val,
        })
    }
}

impl<'a, T> core::ops::Deref for Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.data
    }
}

impl<'a, T> core::ops::DerefMut for Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.data
    }
}

impl<'a, T> Drop for Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    fn drop(&mut self) {
        let _ = BorshCodec::store::<T>(self.account, &self.data);
    }
}

/// Alias so callers can write `BorshAccount<'info, T>` like in Anchor.
pub type BorshAccount<'a, T> = Account<'a, T>;

// Allow treating `BorshAccount` as an `AccountInfo` via `AsRef`.
impl<'a, T> AsRef<arch_program::account::AccountInfo<'a>> for Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    fn as_ref(&self) -> &arch_program::account::AccountInfo<'a> {
        self.account
    }
}
