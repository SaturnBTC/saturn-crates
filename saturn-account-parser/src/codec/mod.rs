//! Serialization codecs used by Saturn account wrappers.
//!
//! The available codecs are:
//! * [`borsh`] – copy-based Borsh (de)serialization.
//! * [`zero_copy`] – zero-copy reinterpretation into Plain-Old-Data structs for
//!   maximum performance on-chain.

pub mod borsh;
pub mod zero_copy;

pub use borsh::{Account, BorshAccount, BorshCodec};
pub use zero_copy::{AccountLoader, ZeroCopyAccount, ZeroCopyCodec};
