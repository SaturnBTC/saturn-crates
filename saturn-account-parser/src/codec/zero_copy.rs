//! Zero-copy codec and account loader.
//!
//! Unlike the Borsh codec, zero-copy access reinterprets the raw account
//! buffer as a Plain-Old-Data (`Pod`) struct without performing any heap
//! allocation or copy.  This provides maximum throughput at the cost of
//! stricter type and alignment requirements.
//!
//! The primary entry points are [`ZeroCopyCodec`] for bare loading/storing and
//! [`AccountLoader`] for an Anchor-style wrapper that
//! tracks borrows at runtime.

use arch_program::account::AccountInfo;
use arch_program::program_error::ProgramError;
use bytemuck::{Pod, Zeroable};
use core::mem::{align_of, size_of};
use std::cell::{Ref, RefMut};

/// Length in bytes of the account discriminator that prefixes every
/// zero-copy account.
const DISCRIMINATOR_LEN: usize = 8;

/// Every zero-copy account type must provide an 8-byte discriminator that
/// uniquely identifies its layout on-chain.  It mirrors Anchorâ€™s convention
/// and will be written as the first eight bytes of the account data.
pub trait Discriminator {
    /// The constant 8-byte discriminator.  Implementations are typically
    /// generated via the upcoming `#[derive(Discriminator)]` procedural macro.
    const DISCRIMINATOR: [u8; 8];
}

/// Zero-copy codec: re-interprets the account data buffer as a `T` without
/// performing any heap allocations or copies.
pub struct ZeroCopyCodec;

impl ZeroCopyCodec {
    /// Copies the shard out of the account into an owned value.
    pub fn load_copy<S>(account: &AccountInfo<'_>) -> Result<S, ProgramError>
    where
        S: Pod + Zeroable + Clone + Discriminator,
    {
        let data = account.try_borrow_data()?;
        if data.len() < DISCRIMINATOR_LEN + size_of::<S>() {
            return Err(ProgramError::InvalidAccountData);
        }
        // Verify discriminator matches.
        if &data[..DISCRIMINATOR_LEN] != &<S as Discriminator>::DISCRIMINATOR
        {
            return Err(ProgramError::InvalidAccountData);
        }

        // SAFETY: bounds checked above + `S` is Pod.
        let bytes = &data[DISCRIMINATOR_LEN..DISCRIMINATOR_LEN + size_of::<S>()];
        Ok(*bytemuck::from_bytes(bytes))
    }

    /// Stores an owned value back into the account's data buffer.
    pub fn store_copy<S>(account: &AccountInfo<'_>, shard: &S) -> Result<(), ProgramError>
    where
        S: Pod + Zeroable + Discriminator,
    {
        let mut data = account.try_borrow_mut_data()?;
        if data.len() < DISCRIMINATOR_LEN + size_of::<S>() {
            return Err(ProgramError::InvalidAccountData);
        }
        if &data[..DISCRIMINATOR_LEN] != &<S as Discriminator>::DISCRIMINATOR
        {
            return Err(ProgramError::InvalidAccountData);
        }
        // SAFETY: same as in `load_copy`.
        let bytes = &mut data[DISCRIMINATOR_LEN..DISCRIMINATOR_LEN + size_of::<S>()];
        bytes.copy_from_slice(bytemuck::bytes_of(shard));
        Ok(())
    }

    /// Returns a mutable reference into the account's data buffer interpreted as `S`.
    pub fn load_mut_ref<'a, S>(account: &'a AccountInfo<'a>) -> Result<RefMut<'a, S>, ProgramError>
    where
        S: Pod + Zeroable + Discriminator + 'static,
    {
        // Disallow mutable access when the account was not marked writable by the caller.
        if !account.is_writable {
            return Err(ProgramError::Custom(
                crate::error::ErrorCode::IncorrectIsWritableFlag.into(),
            ));
        }

        let data = account.try_borrow_mut_data()?;

        if data.len() < DISCRIMINATOR_LEN + size_of::<S>() {
            return Err(ProgramError::InvalidAccountData);
        }

        if &data[..DISCRIMINATOR_LEN] != &<S as Discriminator>::DISCRIMINATOR
        {
            return Err(ProgramError::InvalidAccountData);
        }

        // Ensure proper alignment (after discriminator offset).
        if (data[DISCRIMINATOR_LEN..].as_ptr() as usize) % align_of::<S>() != 0 {
            return Err(ProgramError::InvalidAccountData);
        }

        // SAFETY: alignment + size checks above guarantee safe reinterpretation.
        let ref_mut = RefMut::map(data, |slice| {
            let slice = &mut slice[DISCRIMINATOR_LEN..DISCRIMINATOR_LEN + size_of::<S>()];
            unsafe { &mut *(slice.as_mut_ptr() as *mut S) }
        });

        Ok(ref_mut)
    }

    /// Returns an immutable reference into the account's data buffer.
    pub fn load_ref<'a, S>(account: &'a AccountInfo<'a>) -> Result<Ref<'a, S>, ProgramError>
    where
        S: Pod + Zeroable + Discriminator + 'static,
    {
        let data = account.try_borrow_data()?;

        if data.len() < DISCRIMINATOR_LEN + size_of::<S>() {
            return Err(ProgramError::InvalidAccountData);
        }

        if &data[..DISCRIMINATOR_LEN] != &<S as Discriminator>::DISCRIMINATOR
        {
            return Err(ProgramError::InvalidAccountData);
        }

        // Ensure proper alignment.
        if (data[DISCRIMINATOR_LEN..].as_ptr() as usize) % align_of::<S>() != 0 {
            return Err(ProgramError::InvalidAccountData);
        }

        // SAFETY: same guarantees as `load_mut_ref`, but immutable.
        let ref_imm = Ref::map(data, |slice| {
            let slice = &slice[DISCRIMINATOR_LEN..DISCRIMINATOR_LEN + size_of::<S>()];
            unsafe { &*(slice.as_ptr() as *const S) }
        });
        Ok(ref_imm)
    }
}

// -----------------------------------------------------------------------------
// Anchor-style zero-copy account loader
// -----------------------------------------------------------------------------

#[allow(clippy::module_name_repetitions)]
pub struct AccountLoader<'a, T>
where
    T: Pod + Zeroable + Discriminator + 'static,
{
    account: &'a AccountInfo<'a>,
    _phantom: core::marker::PhantomData<T>,
}

// ---------------- Generic helper methods ----------------
impl<'a, T> AccountLoader<'a, T>
where
    T: Pod + Zeroable + Discriminator + 'static,
{
    /// Creates a new loader wrapping the given account.
    pub fn new(account: &'a AccountInfo<'a>) -> Self {
        Self {
            account,
            _phantom: core::marker::PhantomData,
        }
    }

    /// Immutable borrow of the underlying zero-copy struct.
    pub fn load(&self) -> Result<Ref<'a, T>, ProgramError> {
        ZeroCopyCodec::load_ref::<T>(self.account)
    }

    /// Direct access to the wrapped `AccountInfo`.
    pub fn info(&self) -> &'a AccountInfo<'a> {
        self.account
    }

    /// Mutable borrow of the underlying zero-copy struct.
    pub fn load_mut(&self) -> Result<RefMut<'a, T>, ProgramError> {
        ZeroCopyCodec::load_mut_ref::<T>(self.account)
    }

    /// Initialises a brand-new zero-copy account (resize + zero-fill) and returns a mutable reference to it.
    pub fn load_init(&self) -> Result<RefMut<'a, T>, ProgramError>
    where
        T: Default,
    {
        Self::init_zero_copy_account(self.account)
    }

    /// Allocates and zero-initialises an account for zero-copy usage and
    /// returns a mutable reference to the freshly created struct.
    fn init_zero_copy_account<'info>(
        account_info: &'info AccountInfo<'info>,
    ) -> Result<RefMut<'info, T>, ProgramError>
    where
        T: Default,
    {
        let size = core::mem::size_of::<T>();

        let total_size = size + DISCRIMINATOR_LEN;

        // (Re)allocate the account to the exact size and make it rent-exempt.
        account_info.realloc(total_size, true)?;

        // Write discriminator + zero-fill remainder.
        {
            let mut data = account_info.try_borrow_mut_data()?;
            if data.len() < total_size {
                return Err(ProgramError::InvalidAccountData);
            }

            // write discriminator
            data[..DISCRIMINATOR_LEN]
                .copy_from_slice(&<T as Discriminator>::DISCRIMINATOR);

            // zero-fill struct bytes
            for byte in &mut data[DISCRIMINATOR_LEN..total_size] {
                *byte = 0;
            }
        }

        // Return a mutable zero-copy reference.
        ZeroCopyCodec::load_mut_ref::<T>(account_info)
    }

    /// Adjusts the account's data buffer to exactly `size_of::<T>()` bytes and
    /// keeps any excess lamports in place so the caller does not need to fund
    /// rent again.
    pub fn resize_account<'info>(
        account_info: &'info AccountInfo<'info>,
    ) -> Result<(), ProgramError>
    where
        T: Sized + Default,
    {
        let size = core::mem::size_of::<T>() + DISCRIMINATOR_LEN;
        account_info.realloc(size, true)?;
        Ok(())
    }
}

// Allow treating `AccountLoader` (any mutability) as an `AccountInfo` via `AsRef`.
impl<'a, T> AsRef<arch_program::account::AccountInfo<'a>> for AccountLoader<'a, T>
where
    T: Pod + Zeroable + Discriminator + 'static,
{
    fn as_ref(&self) -> &arch_program::account::AccountInfo<'a> {
        self.account
    }
}
