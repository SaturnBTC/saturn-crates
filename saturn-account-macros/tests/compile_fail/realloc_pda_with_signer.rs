use saturn_account_macros::Accounts;
use saturn_account_parser::codec::Account;

#[derive(Accounts)]
struct ReallocPdaSigner<'info> {
    #[account(mut, signer)]
    user: Account<'info, u64>,
    // PDA account incorrectly marked signer while also realloc
    #[account(mut, realloc, signer, seeds = &[b"seed"], program_id = Pubkey::default(), space = 64)]
    pda_data: Account<'info, u64>,
}

fn main() {}
