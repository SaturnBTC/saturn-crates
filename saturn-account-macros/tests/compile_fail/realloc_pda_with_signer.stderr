error: `realloc` PDA account cannot be marked `signer` - the program signs via seeds
  --> tests/compile_fail/realloc_pda_with_signer.rs:10:5
   |
10 |     pda_data: Account<'info, u64>,
   |     ^^^^^^^^
